File: app/src/main/java/com/helloworld/faceinteract/CameraScanner.java
package com.helloworld.faceinteract;

import android.content.Context;
import android.graphics.*;
import android.hardware.camera2.*;
import android.media.Image;
import android.media.ImageReader;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;
import android.util.Size;
import android.view.Surface;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import java.nio.ByteBuffer;
import java.util.Arrays;

/**
 * Scan images from camera
 */
public class CameraScanner
{
    private EngineManager engineManager;
    private FaceDataManager faceDataManager;
    private SessionManager sessionManager;
    private PhotoScanner photoScanner;
    private CameraCreator cameraCreator;
    private Bitmap bitmap;
    private Thread previewThread;
    private Face extractedFace;

    /**
     * Create {@link CameraScanner}
     * @param context App activity
     * @param surfaceView {@link SurfaceView} to show preview
     */
    CameraScanner(Context context, final SurfaceView surfaceView)
    {
        CameraManager manager = (CameraManager) context.getSystemService(Context.CAMERA_SERVICE);
        //Create camera
        cameraCreator = new CameraCreator(context, manager);
        //Set preview size
        cameraCreator.setPreviewSize(new Size(surfaceView.getMeasuredWidth(), surfaceView.getMeasuredHeight()));
        ImageReader.OnImageAvailableListener imageAvailableListener = new ImageReader.OnImageAvailableListener()
        {
            @Override
            public void onImageAvailable(ImageReader imageReader)
            {
                //When picture is taken...
                cameraCreator.getCamera().close();
                Image image = imageReader.acquireNextImage();
                ByteBuffer buffer = image.getPlanes()[0].getBuffer();
                byte[] data = new byte[buffer.remaining()];
                //Save picture data in this.bitmap
                bitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
            }
        };
        cameraCreator.createImageReader(imageAvailableListener);
        sessionManager = new SessionManager(cameraCreator, surfaceView);
    }

    /**
     * Start preview
     */
    public void start()
    {
        sessionManager.createPreview();
        previewThread = new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                while (sessionManager.getSurfaceHolder() == null)
                {
                    try
                    {
                        Thread.sleep(100);
                    }
                    catch (InterruptedException e)
                    {
                        e.printStackTrace();
                    }
                }
                while (sessionManager.getSurfaceHolder() != null)
                {
                    //How to get preview image???
                    Canvas canvas = sessionManager.getSurfaceHolder().lockCanvas();
                    //If preview image is already in this canvas...
                    if (canvas != null)
                    {
                        Paint paint = new Paint();
                        paint.setARGB(128,255,160,0);
                        paint.setAntiAlias(true);
                        paint.setStrokeWidth(2.0f);
                        paint.setStyle(Paint.Style.STROKE);

                        for (Rect rect : photoScanner.getRectList())
                        {
                            canvas.drawRect(rect, paint);
                        }
                    }
                }
            }
        });
        sessionManager.setSurfaceDestroyedHandler(new Runnable()
        {
            @Override
            public void run()
            {
                try
                {
                    previewThread.join();
                }
                catch (InterruptedException e)
                {
                    e.printStackTrace();
                }
            }
        });
        previewThread.start();
    }

    public void takePicture()
    {
        sessionManager.takePicture();
    }

    public Bitmap getScannedBitmap()
    {
        photoScanner = new PhotoScanner(bitmap);
        photoScanner.setEngineManager(engineManager);
        photoScanner.setFaceDataManager(faceDataManager);
        //Get scanned bitmap by PhotoScanner
        Bitmap result = photoScanner.getScannedBitmap();
        //Save face
        extractedFace = photoScanner.extractFace();
        return result;
    }

    /**
     * Get extracted face after calling getScannedBitmap
     * @return Extracted face
     */
    public Face extractFace()
    {
        return extractedFace;
    }
    public void setEngineManager(EngineManager engineManager)
    {
        this.engineManager = engineManager;
    }

    public void setFaceDataManager(FaceDataManager faceDataManager)
    {
        this.faceDataManager = faceDataManager;
    }
}

/**
 * Helper class for creating camera
 */
class CameraCreator
{
    private CameraManager manager;
    private Handler childHandler, mainHandler;
    private Size previewSize;
    private CameraDevice camera;
    private ImageReader imageReader;

    /**
     * Create {@link CameraCreator}
     * @param context App activity
     * @param cameraManager {@link CameraManager} to create camera
     */
    CameraCreator(Context context, CameraManager cameraManager)
    {
        assert manager != null;
        assert context != null;
        manager = cameraManager;

        //Start background thread
        HandlerThread thread = new HandlerThread("Camera2");
        thread.start();
        childHandler = new Handler(thread.getLooper());
        mainHandler = new Handler(context.getMainLooper());

        try //Open camera
        {
            CameraDevice.StateCallback stateCallback = new CameraDevice.StateCallback()
            {
                @Override
                public void onOpened(CameraDevice cameraDevice)
                {
                    //Save the active camera
                    camera = cameraDevice;
                }

                @Override
                public void onDisconnected(CameraDevice cameraDevice)
                {
                    //Close current camera
                    camera.close();
                    camera = null;
                }

                @Override
                public void onError(CameraDevice cameraDevice, int i)
                {
                    Log.e("CameraCreator", "Failed to open camera : " + i);
                }
            };
            manager.openCamera(
                    Integer.toString(CameraCharacteristics.LENS_FACING_BACK),
                    stateCallback,
                    null
            );
        }
        catch (CameraAccessException e)
        {
            e.printStackTrace();
        }

    }

    /**
     * Create {@link ImageReader} and set listener on image available
     * @param imageAvailableListener Listener on image available
     */
    public void createImageReader(ImageReader.OnImageAvailableListener imageAvailableListener)
    {
        imageReader = ImageReader.newInstance(getPreviewSize().getWidth(), getPreviewSize().getHeight(),
                ImageFormat.NV21, 1);
        imageReader.setOnImageAvailableListener(imageAvailableListener, mainHandler);
    }
    private Size getPreviewSize()
    {
        return previewSize;
    }

    public void setPreviewSize(Size previewSize)
    {
        this.previewSize = previewSize;
    }

    /**
     * Get active camera
     * @return {@link CameraDevice}
     */
    public CameraDevice getCamera()
    {
        return camera;
    }

    public Handler getChildHandler()
    {
        return childHandler;
    }

    /**
     * Get {@link ImageReader} created by createImageReader
     * @return {@link ImageReader}
     */
    public ImageReader getImageReader()
    {
        return imageReader;
    }
}

/**
 * Manage sessions between camera and app
 */
class SessionManager implements SurfaceHolder.Callback
{
    private CameraCreator cameraCreator;
    private SurfaceView surfaceView;
    private SurfaceHolder surfaceHolder;
    private CaptureRequest.Builder previewBuilder;
    private CameraCaptureSession captureSession;
    private Runnable surfaceDestroyedHandler;

    /**
     * Create {@link SessionManager}
     * @param creator {@link CameraCreator} for camera data
     * @param surfaceView {@link SurfaceView} to show preview
     */
    SessionManager(final CameraCreator creator, SurfaceView surfaceView)
    {
        this.surfaceView = surfaceView;
        cameraCreator = creator;
    }

    /**
     * Create preview session
     */
    public void createPreview()
    {
        try
        {
            previewBuilder = cameraCreator.getCamera().createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
            Surface surface = surfaceView.getHolder().getSurface();

            previewBuilder.addTarget(surface);

            cameraCreator.getCamera().createCaptureSession(Arrays.asList(surface, cameraCreator.getImageReader().getSurface()), new CameraCaptureSession.StateCallback()
            {
                @Override
                public void onConfigured(CameraCaptureSession cameraCaptureSession)
                {
                    captureSession = cameraCaptureSession;
                    try
                    {
                        previewBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);
                        //                        previewBuilder.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_ON_AUTO_FLASH);
                        CaptureRequest previewRequest = previewBuilder.build();
                        captureSession.setRepeatingRequest(previewRequest, null, cameraCreator.getChildHandler());
                    }
                    catch (CameraAccessException e) {
                        e.printStackTrace();
                    }
                }

                @Override
                public void onConfigureFailed(CameraCaptureSession cameraCaptureSession)
                {
                    Log.e("SessionManager", "Config Failed");
                }
            }, cameraCreator.getChildHandler());
        }
        catch (CameraAccessException e)
        {
            e.printStackTrace();
        }
    }

    /**
     * Take picture from camera
     */
    public void takePicture()
    {
        CaptureRequest.Builder takePictureRequestBuilder;
        try
        {
            takePictureRequestBuilder = cameraCreator.getCamera().createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE);
            takePictureRequestBuilder.addTarget(cameraCreator.getImageReader().getSurface());
            takePictureRequestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);
            takePictureRequestBuilder.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_ON_AUTO_FLASH);
            CaptureRequest mCaptureRequest = takePictureRequestBuilder.build();
            captureSession.capture(mCaptureRequest, null, cameraCreator.getChildHandler());
        }
        catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void surfaceCreated(SurfaceHolder surfaceHolder)
    {
        this.surfaceHolder = surfaceHolder;
    }

    @Override
    public void surfaceChanged(SurfaceHolder surfaceHolder, int i, int i1, int i2)
    {

    }

    @Override
    public void surfaceDestroyed(SurfaceHolder surfaceHolder)
    {
        this.surfaceHolder = null;
        if (surfaceDestroyedHandler != null)
        {
            surfaceDestroyedHandler.run();
        }
    }

    public SurfaceHolder getSurfaceHolder()
    {
        return surfaceHolder;
    }

    public void setSurfaceDestroyedHandler(Runnable surfaceDestroyedHandler)
    {
        this.surfaceDestroyedHandler = surfaceDestroyedHandler;
    }
}

File: app/src/main/java/com/helloworld/faceinteract/CameraScanner.kt
package com.helloworld.faceinteract

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Paint
import android.hardware.camera2.*
import android.media.ImageReader
import android.os.Handler
import android.os.HandlerThread
import android.util.Log
import android.util.Size
import android.view.SurfaceHolder
import android.view.SurfaceView
import java.util.*
import kotlin.concurrent.thread

/**
 * Scan images from camera
 */
class CameraScanner
/**
 * Create [CameraScanner]
 *
 * @param context     App activity
 * @param surfaceView [SurfaceView] to show preview
 */
internal constructor(context: Context, surfaceView: SurfaceView) {
    private var engineManager: EngineManager? = null
    private var faceDataManager: FaceDataManager? = null
    private val sessionManager: SessionManager
    private val cameraCreator: CameraCreator
    private var bitmap: Bitmap? = null
    private var previewThread: Thread? = null
    private var extractedFace: Face? = null

    //Get scanned bitmap by PhotoScanner
    //Save face
    val scannedBitmap: Bitmap
        get() {
            val photoScanner = PhotoScanner(bitmap!!)
            photoScanner.setEngineManager(engineManager!!)
            photoScanner.setFaceDataManager(faceDataManager!!)
            val result = photoScanner.scannedBitmap
            extractedFace = photoScanner.extractFace()
            return result
        }

    init {
        val manager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
        //Create camera
        cameraCreator = CameraCreator(context, manager)
        //Set preview size
        cameraCreator.previewSize = Size(surfaceView.measuredWidth, surfaceView.measuredHeight)
        val imageAvailableListener = ImageReader.OnImageAvailableListener { imageReader ->
            //When picture is taken...
            cameraCreator.camera.close()
            val image = imageReader.acquireNextImage()
            val buffer = image.planes[0].buffer
            val data = ByteArray(buffer.remaining())
            //Save picture data in this.bitmap
            bitmap = BitmapFactory.decodeByteArray(data, 0, data.size)
        }
        cameraCreator.createImageReader(imageAvailableListener)
        sessionManager = SessionManager(cameraCreator, surfaceView)
    }

    /**
     * Start preview
     */
    fun start() {
        sessionManager.createPreview()
        previewThread = thread(start = true) {
            while (sessionManager.surfaceHolder == null) {
                try {
                    Thread.sleep(100)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
            while (sessionManager.surfaceHolder != null) {
                //How to get preview image???
                val canvas = sessionManager.surfaceHolder!!.lockCanvas()
                if (canvas != null) {

                    val paint = Paint()
                    paint.setARGB(128, 255, 160, 0)
                    paint.isAntiAlias = true
                    paint.strokeWidth = 2.0f
                    paint.style = Paint.Style.STROKE

                    // TODO: get preview image -> scan -> draw rectangles
//                    for (rect in photoScanner!!.getRectList()!!) {
//                        canvas.drawRect(rect, paint)
//                    }
                }
            }
        }
        sessionManager.setSurfaceDestroyedHandler {
            try {
                previewThread!!.join()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    fun takePicture() {
        sessionManager.takePicture()
    }

    /**
     * Get extracted face after calling getScannedBitmap
     *
     * @return Extracted face
     */
    fun extractFace(): Face? {
        return extractedFace
    }

    fun setEngineManager(engineManager: EngineManager) {
        this.engineManager = engineManager
    }

    fun setFaceDataManager(faceDataManager: FaceDataManager) {
        this.faceDataManager = faceDataManager
    }
}

/**
 * Helper class for creating camera
 */
internal class CameraCreator
/**
 * Create [CameraCreator]
 *
 * @param context       App activity
 * @param manager [CameraManager] to create camera
 */
(context: Context?, private val manager: CameraManager?) {
    val childHandler: Handler
    private val mainHandler: Handler
    var previewSize = Size(0, 0)
    /**
     * Get active camera
     *
     * @return [CameraDevice]
     */
    lateinit var camera: CameraDevice
        private set
    /**
     * Get [ImageReader] created by createImageReader
     *
     * @return [ImageReader]
     */
    lateinit var imageReader: ImageReader
        private set

    init {
        assert(manager != null)
        assert(context != null)

        //Start background thread
        val thread = HandlerThread("Camera2")
        thread.start()
        childHandler = Handler(thread.looper)
        mainHandler = Handler(context!!.mainLooper)

        try
        //Open camera
        {
            val stateCallback = object : CameraDevice.StateCallback() {
                override fun onOpened(cameraDevice: CameraDevice) {
                    //Save the active camera
                    camera = cameraDevice
                }

                override fun onDisconnected(cameraDevice: CameraDevice) {
                    //Close current camera
                    camera.close()
                }

                override fun onError(cameraDevice: CameraDevice, i: Int) {
                    Log.e("CameraCreator", "Failed to open camera : " + i)
                }
            }
            manager!!.openCamera(
                    Integer.toString(CameraCharacteristics.LENS_FACING_BACK),
                    stateCallback, null
            )
        } catch (e: CameraAccessException) {
            e.printStackTrace()
        }
    }

    /**
     * Create [ImageReader] and set listener on image available
     *
     * @param imageAvailableListener Listener on image available
     */
    fun createImageReader(imageAvailableListener: ImageReader.OnImageAvailableListener) {
        imageReader = ImageReader.newInstance(previewSize.width, previewSize.height,
                ImageFormat.NV21, 1)
        imageReader.setOnImageAvailableListener(imageAvailableListener, mainHandler)
    }
}

/**
 * Manage sessions between camera and app
 */
internal class SessionManager
/**
 * Create [SessionManager]
 *
 * @param cameraCreator [CameraCreator] for camera data
 * @param surfaceView   [SurfaceView] to show preview
 */
(private val cameraCreator: CameraCreator, private val surfaceView: SurfaceView) : SurfaceHolder.Callback {
    var surfaceHolder: SurfaceHolder? = null
        private set
    private lateinit var previewBuilder: CaptureRequest.Builder
    private lateinit var captureSession: CameraCaptureSession
    private var surfaceDestroyedHandler: (() -> Unit)? = null

    /**
     * Create preview session
     */
    fun createPreview() {
        try {
            previewBuilder = cameraCreator.camera.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)
            val surface = surfaceView.holder.surface

            previewBuilder.addTarget(surface)

            cameraCreator.camera.createCaptureSession(
                    Arrays.asList(surface, cameraCreator.imageReader.surface),
                    object : CameraCaptureSession.StateCallback() {
                override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
                    captureSession = cameraCaptureSession
                    try {
                        previewBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
                        //                        previewBuilder.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_ON_AUTO_FLASH);
                        val previewRequest = previewBuilder.build()
                        captureSession.setRepeatingRequest(previewRequest, null, cameraCreator.childHandler)
                    } catch (e: CameraAccessException) {
                        e.printStackTrace()
                    }

                }

                override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
                    Log.e("SessionManager", "Config Failed")
                }
            }, cameraCreator.childHandler)
        } catch (e: CameraAccessException) {
            e.printStackTrace()
        }

    }

    /**
     * Take picture from camera
     */
    fun takePicture() {
        val takePictureRequestBuilder: CaptureRequest.Builder
        try {
            takePictureRequestBuilder = cameraCreator.camera.createCaptureRequest(CameraDevice.TEMPLATE_STILL_CAPTURE)
            takePictureRequestBuilder.addTarget(cameraCreator.imageReader.surface)
            takePictureRequestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
            takePictureRequestBuilder.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_ON_AUTO_FLASH)
            val mCaptureRequest = takePictureRequestBuilder.build()
            captureSession.capture(mCaptureRequest, null, cameraCreator.childHandler)
        } catch (e: CameraAccessException) {
            e.printStackTrace()
        }

    }

    override fun surfaceCreated(surfaceHolder: SurfaceHolder) {
        this.surfaceHolder = surfaceHolder
    }

    override fun surfaceChanged(surfaceHolder: SurfaceHolder, i: Int, i1: Int, i2: Int) {

    }

    override fun surfaceDestroyed(surfaceHolder: SurfaceHolder) {
        this.surfaceHolder = null
        surfaceDestroyedHandler?.invoke()
    }

    fun setSurfaceDestroyedHandler(surfaceDestroyedHandler: () -> Unit) {
        this.surfaceDestroyedHandler = surfaceDestroyedHandler
    }
}